guava/src/com/google/common/collect/ImmutableSet.java

```diff
@@ -144,37 +145,36 @@
     elements[4] = e5;
     elements[5] = e6;
     System.arraycopy(others, 0, elements, paramCount, others.length);
-    return construct(elements.length, elements);
+    return construct(elements.length, Iterators.forArray(elements));
   }
 
   /**
-   * Constructs an {@code ImmutableSet} from the first {@code n} elements of the specified array. If
-   * {@code k} is the size of the returned {@code ImmutableSet}, then the unique elements of {@code
-   * elements} will be in the first {@code k} positions, and {@code elements[i] == null} for {@code
-   * k <= i < n}.
-   *
-   * <p>This may modify {@code elements}. Additionally, if {@code n == elements.length} and {@code
-   * elements} contains no duplicates, {@code elements} may be used without copying in the returned
-   * {@code ImmutableSet}, in which case it may no longer be modified.
+   * Constructs an {@code ImmutableSet} from the first {@code n} elements of the specified iterable.
+   * If {@code k} is the size of the returned {@code ImmutableSet}, then the unique elements of
+   * {@code elements} will be in the first {@code k} positions, and {@code elements[i] == null} for
+   * {@code k <= i < n}.
    *
    * <p>{@code elements} may contain only values of type {@code E}.
    *
    * @throws NullPointerException if any of the first {@code n} elements of {@code elements} is null
    */
-  private static <E> ImmutableSet<E> construct(int n, Object... elements) {
+  private static <E> ImmutableSet<E> construct(int n, Iterator<Object> elements) {
+    return construct(n, Builder.DEFAULT_INITIAL_CAPACITY, elements);
+  }
+
+  private static <E> ImmutableSet<E> construct(int n, int expectedSize, Iterator<Object> elements) {
```

Comment:
I'd prefer for this to be `Iterator<E>` and you have to cast the result of `toArray()` instead of the other way around.